sync3/handler/ensure_polling.go

```diff
@@ -93,6 +109,38 @@ func (p *EnsurePoller) OnInitialSyncComplete(payload *pubsub.V2InitialSyncComple
 	close(ch)
 }
 
+func (p *EnsurePoller) OnTokenExpired(payload *pubsub.V2ExpiredToken) {
+	pid := sync2.PollerID{UserID: payload.UserID, DeviceID: payload.DeviceID}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	pending, exists := p.pendingPolls[pid]
+	if !exists {
+		// We weren't tracking the state of this poller, so we have nothing to clean up.
+		return
+	}
+	// There's a potential race here. Suppose that
+	//
+	//  - a v3 request arrives for a brand-new device
+	//  - homeserver recognises the access token in the /whoami call
+	//  - we call EnsurePolling for that token
+	//  - the poller makes an initial sync and learns that the token has expired
+	//
+	// If we close the channel, the EnsurePolling call that's blocking on it will
+	// continue as-if there was initial sync data to receive. But there isn't.
```

Comment:
How is this a race exactly? We're just unblocking the v3 request on expired tokens.

Reply:
Maybe not a race; this just came to mind as a potentially bad outcome. My concern was that the proxy might assume that when the v3 request is unblocked it has some kind of sync v2 data persisted to DB. But I suppose it has to anticipate and handle poller failures anyway?